### PR TITLE
Multi Role Instance Role field

### PR DESCRIPTION
**Pull request by ANeumann82**
Summary:
- Add Role field to Instances
- Add Migration to set mesos_role to all instances without role

JIRA issues: MARATHON-8637

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala [hunk 1]
```diff
@@ -28,7 +28,8 @@ case class Instance(
     state: InstanceState,
     tasksMap: Map[Task.Id, Task],
     runSpec: RunSpec,
-    reservation: Option[Reservation]) extends Placed {
+    reservation: Option[Reservation],
+    role: String) extends Placed {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: I think we should use `type Role = String` as in #6961

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala [hunk 2]
```diff
@@ -418,4 +420,9 @@ object Instance {
       Json.toJson(stringToTask)
     }
   )
+
+  /**
+    * This is the default role that mesos uses when no role is specified by marathon.
+    */
+  val defaultMesosRole = "TBD"
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: Why is this required?

File: src/main/scala/mesosphere/marathon/state/Instance.scala [hunk 3]
```diff
@@ -20,7 +20,8 @@ case class Instance(
     state: InstanceState,
     tasksMap: Map[Task.Id, Task],
     runSpecVersion: Timestamp,
-    reservation: Option[Reservation]) extends MarathonState[Protos.Json, Instance] {
+    reservation: Option[Reservation],
+    role: String) extends MarathonState[Protos.Json, Instance] {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/Instance.scala]: This should be `Option[String]` as it is the storrge model.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala [hunk 4]
```diff
@@ -104,7 +104,7 @@ object MigrationTo17 extends StrictLogging {
         val reservation = rawReservation.map { raw =>
           raw.as[Reservation](InstanceMigration.legacyReservationReads(tasksMap, instanceId))
         }
-        new Instance(instanceId, Some(agentInfo), state, tasksMap, runSpecVersion, reservation)
+        new Instance(instanceId, Some(agentInfo), state, tasksMap, runSpecVersion, reservation, null)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala]: Please use `None` instead for the storage model.

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala (hunk 1 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: That's actually a pretty good idea, but then we might want to use that in the AppDefinition/PodDefinition as well, right? Where would be a good place to define the `type Role = String`?

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala (hunk 2 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: Left over from earlier code. Replaced it with "*" in the tests, wasn't used anywhere else

File: src/main/scala/mesosphere/marathon/state/Instance.scala (hunk 3 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: Why though? It might make the migration a little bit easier, but we'd have to take the optionality into account when converting from storage state to core conversion.

Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: Changed it.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala (hunk 4 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala]: Done

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala (hunk 1 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: Done

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo19100.scala [hunk 5]
```diff
@@ -7,44 +7,91 @@ import java.time.OffsetDateTime
...
     logger.info("Starting migration to 1.9.100")
     await(MigrationTo19100.migrateApps(defaultMesosRole, persistenceStore))
     await(MigrationTo19100.migratePods(defaultMesosRole, persistenceStore))
+    await(InstanceMigration.migrateInstances(instanceRepository, persistenceStore, instanceMigrationFlow))
   }
+
+  /**
+    * Read format for old instance without reservation id.
+    */
+  val instanceJsonReads19100: Reads[Instance] = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo19100.scala]:
```suggestion
  val instanceJsonReads18200: Reads[Instance] = {
```

File: src/main/scala/mesosphere/marathon/state/Instance.scala [hunk 6]
```diff
@@ -41,7 +42,10 @@ case class Instance(
     * @param runSpec The run spec belonging to this instance.
     * @return This instance in core model.
     */
-  def toCoreInstance(runSpec: RunSpec) = CoreInstance(instanceId, agentInfo, state, tasksMap, runSpec, reservation)
+  def toCoreInstance(runSpec: RunSpec) = {
+    require(role.isDefined, "Stored Instances must have a defined role")
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/Instance.scala]:
```suggestion
    require(role.isDefined, s"BUG! Stored $instanceId has no role defined which should be covered by migration.")
```

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      val migration = f.migration()
+
+      When("they are run through the migration flow")
+      val updatedInstances = instances.via(migration.instanceMigrationFlow).runWith(Sink.seq).futureValue
+
+      Then("only one instance has been migrated")
+      updatedInstances should have size (1)
+      updatedInstances.head.instanceId should be(instanceId1)
+      updatedInstances.head.role.isDefined should be(true)
+      updatedInstances.head.role.get should be(mesosDefaultRole)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala]:
```suggestion
      updatedInstances.head.role.value should be(mesosDefaultRole)
```

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+      val migration = f.migration()
+
+      When("they are run through the migration flow")
+      val updatedInstances = instances.via(migration.instanceMigrationFlow).runWith(Sink.seq).futureValue
+
+      Then("only one instance has been migrated")
+      updatedInstances should have size (1)
+      updatedInstances.head.instanceId should be(instanceId1)
+      updatedInstances.head.role.isDefined should be(true)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala]:
```suggestion
```

Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala]: ScalaTest `.value` should cover this.

**Comment by jeschkies** [comment]
I remved the `git clean -fdx` again. It seems to me that it is not executed as sudo.

File: src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala [hunk 9]
```diff
@@ -49,13 +49,13 @@ class InstanceOpFactoryImpl(
...
-      case scheduledInstance @ Instance(_, _, _, _, app: AppDefinition, _) =>
+      case scheduledInstance @ Instance(_, _, _, _, app: AppDefinition, _, _) =>
         if (app.isResident) inferForResidents(request, scheduledInstance)
         else inferNormalTaskOp(app, request.instances, request.offer, request.localRegion, scheduledInstance)
-      case scheduledInstance @ Instance(_, _, _, _, pod: PodDefinition, _) =>
+      case scheduledInstance @ Instance(_, _, _, _, pod: PodDefinition, _, _) =>
         if (pod.isResident) inferForResidents(request, scheduledInstance)
         else inferPodInstanceOp(pod, request.instances, request.offer, request.localRegion, scheduledInstance)
-      case Instance(_, _, _, _, runSpec, _) =>
+      case Instance(_, _, _, _, runSpec, _, _) =>
```
[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala]: Yeah, this code abused extractors.

File: src/main/scala/mesosphere/marathon/state/Group.scala [hunk 10]
```diff
@@ -127,6 +127,7 @@ class Group(
 
 object Group extends StrictLogging {
   type GroupKey = PathId
+  type Role = String
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: Hmm, Role? Here?

Comment (timcharper) [reply src/main/scala/mesosphere/marathon/state/Group.scala]: Maybe we should do a package object and put it as a peer to Group

File: src/main/scala/mesosphere/marathon/state/Instance.scala [hunk 11]
```diff
@@ -41,7 +42,10 @@ case class Instance(
     * @param runSpec The run spec belonging to this instance.
     * @return This instance in core model.
     */
-  def toCoreInstance(runSpec: RunSpec) = CoreInstance(instanceId, agentInfo, state, tasksMap, runSpec, reservation)
+  def toCoreInstance(runSpec: RunSpec) = {
+    require(role.isDefined, s"BUG! Stored $instanceId has no role defined which should be covered by migration.")
```
[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: Yes. Role should be non-optional. Thank you.

[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: I tend to disagree. This is not internal `core.instance.Instance` but storage `state.Instance` model. It is used for migrations and loading only and is similar to `StoredGroup`. I think it's a valid pattern. We could start cleaning up the different storage models JSON, Protobuf, RAML, `state.Instance` and `storage.StoredGroup` and just have one. However, if we eg would go with Protobuf it would also just have the role as an optional field if I'm not mistaken.

File: src/main/scala/mesosphere/marathon/core/instance/Instance.scala [hunk 12]
```diff
@@ -28,7 +29,8 @@ case class Instance(
     state: InstanceState,
     tasksMap: Map[Task.Id, Task],
     runSpec: RunSpec,
-    reservation: Option[Reservation]) extends Placed {
+    reservation: Option[Reservation],
+    role: Role) extends Placed {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: Could you add a comment to the instance? Eg

```scala
/**
 * Internal state of a instantiated [[RunSpec]], ie instance of an [[AppDefinition]]
 * or [[PodDefinition]].
 * ...
 * @param role The Mesos role the instance was launched this. It can differ
 *             from the run spec role. When a resident instance is updated and
 *             restarted we keep the old role to maintain the data. For instance
 *             one might want to update the Cassandra nodes from v1 to v2 but
 *             loose terabytes of data. All old instances would keep running with
 *             role `slave_public` even though the app definition for v2 as role
 *             `prod` and all new instance are launched with role `pod`. This
 *             inconsistency will allow for a smooth migration of persistent apps.
 */
case class Instance(...
    role: Role
```

Comment (timcharper) [reply src/main/scala/mesosphere/marathon/core/instance/Instance.scala]: (Karsten, I hope you don't mind, I edited your comment and re-worded the comment about the instance parameter, shooting for more brevity)

File: src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala]: Nice, done

File: tests/integration/src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 13]
```diff
@@ -215,8 +215,10 @@ class UpgradeIntegrationTest extends AkkaIntegrationTest with MesosClusterTest w
       (marathonCurrent.client.info.entityJson \ "version").as[String] should be(BuildInfo.version.toString)
 
       Then("All apps from 1.4.9 and 1.5.15 are still running")
-      marathonCurrent.client.tasks(app_149.id.toPath).value should contain theSameElementsAs (originalApp149Tasks)
-      marathonCurrent.client.tasks(app_1515.id.toPath).value should contain theSameElementsAs (originalApp1515Tasks)
+      val originalApp149TasksWithRole = originalApp149Tasks.map(t => t.copy(role = Some("foo")))
+      val originalApp1515TasksWithRole = originalApp1515Tasks.map(t => t.copy(role = Some("foo")))
+      marathonCurrent.client.tasks(app_149.id.toPath).value should contain theSameElementsAs (originalApp149TasksWithRole)
```
Comment (jeschkies) [inline tests/integration/src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: Hm, why not just compare `marathonCurrent.client.tasks(app_149.id.toPath).value.map(_.id) should contain theSameElementsAs (originalApp149TaskIds)`?

Comment (ANeumann82) [reply tests/integration/src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: Yeah, comparing IDs is better.

File: tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala [hunk 14]
```diff
@@ -37,13 +41,81 @@ class MultiRoleIntegrationTest extends AkkaIntegrationTest with EmbeddedMarathon
...
+
+      waitForTasks(PathId(appInDev.id), 1) //make sure the app has restarted
+
+      Given("The Update is done")
+      var taskListResult = marathon.tasks(appId)
+
+      Then("One Instance should have the old role, and one the new role")
+      var taskList = taskListResult.value
+
+      taskList.size should be(1)
```
Comment (jeschkies) [inline tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala]:
```suggestion
      taskList should have size(1)
```

File: tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala [hunk 15]
```diff
@@ -37,13 +41,81 @@ class MultiRoleIntegrationTest extends AkkaIntegrationTest with EmbeddedMarathon
...
+
+      waitForTasks(PathId(appInDev.id), 2) // make sure both apps have started
+
+      Given("The Update is done")
+      taskListResult = marathon.tasks(appId)
+
+      Then("One Instance should have the old role, and one the new role")
+      taskList = taskListResult.value
+
+      taskList.size should be(2)
```
Comment (jeschkies) [inline tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala]:
```suggestion
      taskList should have size(2)
```

File: src/main/scala/mesosphere/marathon/state/Instance.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: We have protobuf, then we have this, then we have internal state?

Why not migrate it directly on the protobuf?

[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: I'd prefer to not rely on secondary models for migration since they could evolve in backwards incompatible ways

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+    "only update instances with no role" in {
+
+      Given("an instance with and one without role")
+      val f = new Fixture()
+      val instanceId1 = Instance.Id.forRunSpec(PathId("/app"))
+      val instanceId2 = Instance.Id.forRunSpec(PathId("/app2"))
+
+      val instances = Source(List(
+        f.legacyInstanceJson(instanceId1),
```
Comment (timcharper) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala]: Is there a really good reason why we don't migrate on the protobuf directly? I'd love to understand the reasoning here better.

Comment (timcharper) [reply src/test/scala/mesosphere/marathon/storage/migration/MigrationTo19100Test.scala]: ohhhhhhhhh

We store JSON, not protobuf, for our instances. 🤦‍♂ as you were.

File: src/main/scala/mesosphere/marathon/state/Instance.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/state/Instance.scala]: 🤦‍♂ issue is we don't use protobuf for instances. While I'm inclined to think that we should, that's orthogonal to this change.
